internal/daemon/api_files_test.go

```diff
@@ -854,31 +851,28 @@ func (s *filesSuite) TestWriteUserGroupMocked(c *C) {
 		}
 		return nil
 	}
-	var lookupUserCalls []string
-	lookupUser = func(name string) (*user.User, error) {
-		lookupUserCalls = append(lookupUserCalls, name)
-		return &user.User{Uid: "56"}, nil
-	}
-	var lookupGroupCalls []string
-	lookupGroup = func(name string) (*user.Group, error) {
-		lookupGroupCalls = append(lookupGroupCalls, name)
-		return &user.Group{Gid: "78"}, nil
+	normalizeUidGid = func(uid, gid *int, username, group string) (*int, *int, error) {
+		if uid != nil {
+			return uid, gid, nil
+		}
+		if username == "" {
+			return nil, nil, nil
+		}
+		c.Check(username, Equals, "USER")
+		c.Check(group, Equals, "GROUP")
+		u, g := 56, 78
+		return &u, &g, nil
 	}
 	defer func() {
 		atomicWriteChown = osutil.AtomicWriteChown
-		lookupUser = user.Lookup
-		lookupGroup = user.LookupGroup
+		normalizeUidGid = osutil.NormalizeUidGid
```

Comment:
FWIW, I feel like this has actually turned into a bit of an anti-pattern in the Juju code base. These become brittle, and messy with race conditions. It is global state that you are mutating, etc. I much prefer having Objects that you can inject new dependencies into.

Reply:
I think I'm going to leave this as is. It just doesn't seem great for the caller to have to change a simple global function call like `osutil.NormalizeUidGid()` into this:

```go
utils := osutil.OSUtil{LookupUser: user.Lookup, LookupGroup: user.LookupGroup}
utils.NormalizeUidGid()

# or, if a nil default means "use the real OS function" you could do:
utils := osutil.OSUtil{}
utils.NormalizeUidGid()
```

Seeing these are global / singleton functions I'm okay with having globals for their dependencies. Go runs package-level tests serially by default, so race conditions aren't an issue here.

We can always revisit this later.

Reply:
Agree with Ben here, and we also need to keep in mind that this code base is inherited from snapd, not juju. There's a pattern well established already, so we should either respect it or change it all at once at some point, if we ever find a reason to.